File: 12_passage_pathing/passage_pathing.py
from typing import List, Set, Tuple

START_NOTATION = 'start'
END_NOTATION = 'end'


def get_inputs(filename: str) -> Tuple[List[List[int]], List[str]]:
    with open(filename) as f:
        lines = f.readlines()
    points = set()
    for line in lines:
        two_points = line.strip().split('-')
        for point in two_points:
            if point not in points:
                points.add(point)
    point_list = list(points)
    vertices = [[0 for _ in range(len(point_list))]
                for _ in range(len(point_list))]
    for line in lines:
        two_points = line.strip().split('-')
        vertices[point_list.index(two_points[0])
                 ][point_list.index(two_points[1])] = 1
        vertices[point_list.index(two_points[1])
                 ][point_list.index(two_points[0])] = 1
    return vertices, point_list


def get_num_paths(vertices: List[List[int]], points: List[str], point: str, visited: Set) -> int:
    if point == END_NOTATION:
        return 1
    idx = points.index(point)
    if point.islower():
        visited.add(point)
    num_paths = 0
    for i in range(len(vertices[idx])):
        if vertices[idx][i] == 1 and points[i] not in visited:
            new_visited = visited.copy()
            num_paths += get_num_paths(vertices,
                                       points, points[i], new_visited)
    return num_paths


def get_num_paths_pro(vertices: List[List[int]], points: List[str], point: str, visited: Set, used: bool) -> int:
    if point == END_NOTATION:
        return 1
    idx = points.index(point)
    if point.islower():
        visited.add(point)
    num_paths = 0
    for i in range(len(vertices[idx])):
        if vertices[idx][i] == 1:
            if points[i] not in visited:
                new_visited = visited.copy()
                num_paths += get_num_paths_pro(vertices,
                                               points, points[i], new_visited, used)
            elif points[i] in visited and not used \
                    and points[i] != START_NOTATION and points[i] != END_NOTATION:
                new_visited = visited.copy()
                num_paths += get_num_paths_pro(vertices,
                                               points, points[i], new_visited, True)
    return num_paths


if __name__ == '__main__':
    vertices, points = get_inputs('input.txt')
    for line in vertices:
        print(line)
    print(points)
    print(get_num_paths(vertices, points, 'start', set()))
    print(get_num_paths_pro(vertices, points, 'start', set(), False))

File: 12_passage_pathing/passage_pathing.go
package main

import (
	"fmt"
	"os"
	"strings"
)

var (
	startNotation string = "start"
	endNotation   string = "end"
)

func indexOf(s []string, e string) int {
	for i, a := range s {
		if a == e {
			return i
		}
	}
	return -1
}

func getInputs(filename string) ([][]int, []string, error) {
	data, err := os.ReadFile(filename)
	if err != nil {
		return nil, nil, err
	}
	lines := strings.Split(string(data), "\n")
	pointMap := make(map[string]bool)
	for _, line := range lines {
		line = strings.TrimSpace(line)
		twoPoints := strings.Split(line, "-")
		for _, point := range twoPoints {
			if _, ok := pointMap[point]; !ok {
				pointMap[point] = true
			}
		}
	}
	points := make([]string, 0, len(pointMap))
	for point := range pointMap {
		points = append(points, point)
	}
	vertices := make([][]int, len(points))
	for i := range vertices {
		vertices[i] = make([]int, len(points))
	}
	for _, line := range lines {
		line = strings.TrimSpace(line)
		twoPoints := strings.Split(line, "-")
		vertices[indexOf(points, twoPoints[0])][indexOf(points, twoPoints[1])] = 1
		vertices[indexOf(points, twoPoints[1])][indexOf(points, twoPoints[0])] = 1
	}
	return vertices, points, nil
}

func getNumPaths(vertices [][]int, points []string, point string,
	visited map[string]bool) int {
	if point == endNotation {
		return 1
	}
	if strings.ToLower(point) == point {
		visited[point] = true
	}
	numPaths := 0
	idx := indexOf(points, point)
	for i := range vertices[idx] {
		_, ok := visited[points[i]]
		if vertices[idx][i] == 1 && !ok {
			newVisited := make(map[string]bool)
			for k, v := range visited {
				newVisited[k] = v
			}
			numPaths += getNumPaths(vertices, points, points[i], newVisited)
		}
	}
	return numPaths
}

func getNumPathsPro(vertices [][]int, points []string, point string,
	visited map[string]bool, used bool) int {
	if point == endNotation {
		return 1
	}
	if strings.ToLower(point) == point {
		visited[point] = true
	}
	numPaths := 0
	idx := indexOf(points, point)
	for i := range vertices[idx] {
		if vertices[idx][i] == 0 {
			continue
		}
		_, ok := visited[points[i]]
		newVisited := make(map[string]bool)
		for k, v := range visited {
			newVisited[k] = v
		}
		if !ok {
			numPaths += getNumPathsPro(vertices, points, points[i], newVisited, used)
		} else if !used && points[i] != startNotation && points[i] != endNotation {
			numPaths += getNumPathsPro(vertices, points, points[i], newVisited, true)
		}
	}
	return numPaths
}

func main() {
	vertices, points, err := getInputs("input.txt")
	if err != nil {
		panic(err)
	}
	for _, line := range vertices {
		fmt.Println(line)
	}
	fmt.Println(points)
	fmt.Printf("Number of paths %d\n", getNumPaths(vertices, points,
		startNotation, make(map[string]bool)))
	fmt.Printf("Number of paths %d\n", getNumPathsPro(vertices, points,
		startNotation, make(map[string]bool), false))
}
